crates/pipeline-manager/src/compiler/main.rs

```diff
@@ -0,0 +1,268 @@
+use crate::api::error::ApiError;
+use crate::compiler::error::CompilerError;
+use crate::compiler::rust_compiler::{
+    perform_rust_compilation, rust_compiler_task, RustCompilationError,
+};
+use crate::compiler::sql_compiler::{
+    perform_sql_compilation, sql_compiler_task, SqlCompilationError,
+};
+use crate::compiler::util::recreate_dir;
+use crate::config::{CommonConfig, CompilerConfig};
+use crate::db::storage_postgres::StoragePostgres;
+use crate::db::types::common::Version;
+use crate::db::types::pipeline::PipelineId;
+use crate::db::types::program::ProgramConfig;
+use crate::db::types::tenant::TenantId;
+use crate::error::ManagerError;
+use crate::probe::Probe;
+use actix_files::NamedFile;
+use actix_web::{get, web, HttpRequest, HttpServer, Responder};
+use futures_util::join;
+use log::{error, info};
+use std::str::FromStr;
+use std::sync::Arc;
+use tokio::{spawn, sync::Mutex};
+use uuid::Uuid;
+
+/// Decodes the URL encoded parameter value as a string.
+fn decode_url_encoded_parameter(
+    param: &'static str,
+    value: Option<&str>,
+) -> Result<String, ManagerError> {
+    match value {
+        None => Err(ManagerError::from(ApiError::MissingUrlEncodedParam {
+            param,
+        })),
+        Some(checksum) => Ok(checksum.to_string()),
+    }
+}
+
+/// Validates the provided string is a hex-encoded SHA256 checksum.
+fn validate_is_sha256_checksum(s: &str) -> Result<(), ManagerError> {
+    if s.len() != 64 {
+        return Err(ApiError::InvalidChecksumParam {
+            value: s.to_string(),
+            error: format!("{} characters long instead of expected 64", s.len()),
+        }
+        .into());
+    };
+    for c in s.chars() {
+        if !c.is_ascii_digit()
+            && c != 'a'
+            && c != 'b'
+            && c != 'c'
+            && c != 'd'
+            && c != 'e'
+            && c != 'f'
+        {
+            return Err(ApiError::InvalidChecksumParam {
+                value: s.to_string(),
+                error: format!("character '{c}' is not hexadecimal (0-9, a-f)"),
+            }
+            .into());
+        }
+    }
+    Ok(())
+}
+
+/// Retrieves the binary executable.
+#[get("/binary/{pipeline_id}/{program_version}/{source_checksum}/{integrity_checksum}")]
+async fn get_binary(
```

Comment:
seems strange that client has to provide a checksum to get a binary?

Reply:
It's the pipeline itself retrieving it, the main motivation to add the checksum is add a triple check that it gets exactly what it wants (both source checksum and integrity checksum matching). It's auto-generated and not user-facing.